File: Milla/Stack/DailyTemperatures.java
package Milla.Stack;

import java.util.Arrays;
import java.util.Stack;

public class DailyTemperatures {
    /*
    * Edge case
    * 1) if all temperatures are the same, do I return 0?
    * 2) Can the temperature array be empty?
    *
    * Idea
    * using stack for the waiting list
    *
    * Plan
    *
    *
    *
    * TC : O(n). Because I iterate through the array once.
    * SC : I'll use answer, stack it means O(n) space each structure and
    * the total SC is O(n) + O(n) so is going to O(n).
    * */
//웨이팅리스트를 스택으로 저장
    public static void main(String[] args) {
        int a[] = {30,38,30,36,35,40,28};
        System.out.println(Arrays.toString(dailyTemperatures(a)));
    }
    public static int[] dailyTemperatures(int[] temperatures) {
        int[] ans = new int[temperatures.length];
        Stack<Integer> stack = new Stack<>();

        //I'll loop through each day
        for (int i = 0; i < temperatures.length ; i++){
            /*
            * for each day, I'll check if the stack is empty
            * */
            while(!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]){
                int prev = stack.pop();
                ans[prev] = i - prev;
            }
            stack.push(i);
        }
        return ans;
    }
}

File: Joy/Stack DS/daily-temperatures.py
'''
Explore
input: array[int] - temperatures with each day represents the daily temperatures
output: array[int] - answer[i] is the number of days you have to wait to get it warmer

Clarifications
- Can the array be empty? No. 
- If the array have all the same temperature, do I return [0,0,0...]? Yes

Test Cases
[73,74,75,71,69,72,76,73] -> [1,1,4,2,1,1,0,0]
[30, 20, 10] -> [0,0,0]
[10] -> [0]
[10,10,10] -> [0,0,0]

Brainstrom
- Brute Force: use nested loop to compare each day O(N ^2)
- Optimal Approach: Use stack. We will only store the previous index

Plan
1. Declare ans array filled with 0s and stack
2. Iterate through teh tempeartues array
    2-1. If stack and the top of the stack is less than the current temperature (high temperature found!)
        2-2. pop the stack, and store the value to the previous  index
        2-3. Update the ans for the current position. (i - previous index)
    2-4. Add the index to the stack
    2-5. Return the answer array

=> TC: O(N), SC: O(N)
'''

def dailyTemperatures(temperatures):
    ans = [ 0 for _ in range(len(temperatures))]
    stack = []

    for i, n in enumerate(temperatures):
        while stack and temperatures[stack[-1]] < n:
            prev_index = stack.pop()
            ans[prev_index] = i - prev_index
        stack.append(i)
    return ans


# normal case
print(dailyTemperatures([73,74,75,71,69,72,76,73]) == [1,1,4,2,1,1,0,0])
print(dailyTemperatures([30,20,10]) == [0,0,0])

# edge case
print(dailyTemperatures([10]) == [0])
print(dailyTemperatures([10,10,10]) == [0,0,0])